Add BerConverter tests

hughbe opened the pull request:
There's an issue somewhere (that I can't seem to find for the life of me) that this contributes to

hughbe commented on src/System.DirectoryServices.Protocols/tests/BerConverterTests.cs:
```diff
@@ -0,0 +1,174 @@
+{
+    public class BerConverterTests
+    {
+        public static IEnumerable<object[]> Encode_TestData()
```
I would have liked to share this data with Decode (to roundtrip), but the formats are different for encode and decode. Even when I adjust that, many bytes don't roundtrip

stephentoub commented on src/System.DirectoryServices.Protocols/tests/BerConverterTests.cs:
```diff
@@ -0,0 +1,174 @@
+            catch
+            {
+                Console.WriteLine($"Expected: {string.Join(", ", expected)}");
+                Console.WriteLine($"Actual: {string.Join(", ", actual)}");
```
It'd be better to put this data into the exception message.

hughbe commented:
I've addressed your PR feedback and reverted the change. I introduced a helper `AssertExtensions.Equal(byte[], byte[])`

An example error message:

```
System.DirectoryServices.Protocols.Tests.BerConverterTests.Encode_Objects_ReturnsExpected(format: \"VVVV\", values
  : [null, [[...], null], [[...]], []], expected: [4, 4, 220, 1, 2, ...]) [FAIL]
        Exception of type 'Xunit.Sdk.AssertActualExpectedException' was thrown.
        Expected: 4, 4, 220, 1, 2, 3, 4, 0, 4, 0
        Actual:   4, 4, 0, 1, 2, 3, 4, 0, 4, 0
        Stack Trace:
              at System.AssertExtensions.Equal(Byte[] expected, Byte[] actual)
              at System.DirectoryServices.Protocols.Tests.BerConverterTests.Encode_Objects_ReturnsExpected(String forma
  t, Object[] values, Byte[] expected)
```

hughbe commented:
Test NETFX x86 Release Build

danmosemsft commented:
@dotnet-bot test Linux x64 Release Build  please (unrelated hangs)

hughbe commented:
I've just pushed a change to make BerConvert static. Previously this was sealed with a private constructor (like the Brushes etc. in System.Drawing))

stephentoub commented:
@dotnet/dnceng, more fetch failures :frowning:, e.g.
```
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from https://github.com/dotnet/corefx
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:809)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1076)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1107)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:109)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:130)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:59)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:233)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
```

stephentoub commented:
@dotnet-bot test this please

hughbe commented:
@dotnet-bot test this please

hughbe commented:
Test Windows x64 Debug Build

hughbe commented:
Test Windows x64 Debug Build

mmitche commented:
@hughbe What issue were you seeing that needed retries?

mmitche commented:
@dotnet/dnceng for visibility

hughbe commented:
Test failures in windows Nano that were fixed by updating coreclr and test failures in release mode that were fixed by disabling some tests before
